demos/custom-tabs-example-app/src/main/java/org/chromium/customtabsdemos/EngagementSignalsActivity.java

```diff
@@ -0,0 +1,193 @@
+    private TextView mTextSessionEnd;
+    private TextView mTextNavigation;
+
+    private ServiceConnection mConnection;
+    private CustomTabsClient mClient;
+    private CustomTabsSession mCustomTabsSession;
+
+    private ServiceConnectionCallback mServiceConnectionCallback = new ServiceConnectionCallback() {
+        @Override
+        public void onServiceConnected(CustomTabsClient client) {
+            mClient = client;
+            mCustomTabsSession = mClient.newSession(mCustomTabsCallback);
+            try {
+                boolean engagementSignalsApiAvailable = mCustomTabsSession.isEngagementSignalsApiAvailable(Bundle.EMPTY);
+                if (!engagementSignalsApiAvailable) {
+                    Log.d(TAG, "CustomTab Engagement signals not available, make sure to use the " +
+                            "latest Chrome version and enable via chrome://flags/#cct-real-time-engagement-signals");
+                    return;
+                }
+                boolean signalsCallback = mCustomTabsSession.setEngagementSignalsCallback(mEngagementSignalsCallback, Bundle.EMPTY);
+                if (!signalsCallback) {
+                    Log.w(TAG, "Could not set EngagementSignalsCallback");
+                }
+            } catch (RemoteException e) {
+                Log.w(TAG, "The Service died while responding to the request.", e);
+            } catch (UnsupportedOperationException e) {
+                Log.w(TAG, "Engagement Signals API isn't supported by the browser.", e);
+            }
+        }
+
+        @Override
+        public void onServiceDisconnected() {
+            mClient = null;
+            mConnection = null;
+            mCustomTabsSession = null;
+        }
+    };
+
+    private EngagementSignalsCallback mEngagementSignalsCallback = new EngagementSignalsCallback() {
+        @Override
+        public void onVerticalScrollEvent(boolean isDirectionUp, @NonNull Bundle extras) {
+            Log.d(TAG, "onVerticalScrollEvent (isDirectionUp=" + isDirectionUp + ')');
+            mTextVerticalScroll.setText("vertical scroll " + (isDirectionUp ? "UP️" : "DOWN"));
+        }
+
+        @Override
+        public void onGreatestScrollPercentageIncreased(int scrollPercentage, @NonNull Bundle extras) {
+            Log.d(TAG, "scroll percentage: " + scrollPercentage + "%");
+            mTextGreatestPercentage.setText("scroll percentage: " + scrollPercentage + "%");
+        }
+
+        @Override
+        public void onSessionEnded(boolean didUserInteract, @NonNull Bundle extras) {
+            Log.d(TAG, "onSessionEnded (didUserInteract=" + didUserInteract + ')');
+            mTextSessionEnd.setText(didUserInteract ? "session ended with user interaction" : "session ended without user interaction");
+        }
+    };
+
+    private CustomTabsCallback mCustomTabsCallback = new CustomTabsCallback() {
+        @Override
+        public void onNavigationEvent(int navigationEvent, @Nullable Bundle extras) {
```

Comment:
Could you extract this code and [this code][1] to a common method somewhere?

[1]: https://github.com/GoogleChrome/android-browser-helper/blob/d92ca82209082f00567cdd3e73b4c0ff1d720df7/demos/custom-tabs-navigation-callbacks/src/main/java/com/google/androidbrowserhelper/demos/customtabssession/MainActivity.java#L120

Reply:
For prod code I'd agree, but in this case I prefer the code to be present here as documentation. Also added resetting the text inputs on NAVIGATION_STARTED

Reply:
Fair enough.